File: cypher-to-sql/src/main/java/ingraph/compiler/sql/ExportStep.java
package ingraph.compiler.sql;

import org.neo4j.driver.v1.Record;
import org.neo4j.driver.v1.Session;
import org.neo4j.driver.v1.StatementResult;
import org.neo4j.driver.v1.Transaction;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.Collections;

public class ExportStep {
	private String exportCypherQuery, tableName;

	public ExportStep(String exportCypherQuery, String tableName) {
		this.exportCypherQuery = exportCypherQuery;
		this.tableName = tableName;
	}

	public void exportToTable(Session cypherSession, Connection sqlConnection) throws SQLException {
		try (Transaction cypherTransaction = cypherSession.beginTransaction()) {
			StatementResult cypherResult = cypherTransaction.run(exportCypherQuery);

			int keyNumber = cypherResult.keys().size();
			String insertQueryString = "INSERT INTO " + tableName + " VALUES (" + String.join(", ", Collections.nCopies(keyNumber, "?")) + ")";

			try (PreparedStatement insertStatement = sqlConnection.prepareStatement(insertQueryString)) {
				for (Record cypherRecord : (Iterable<Record>) () -> cypherResult) {
					for (int keyIndex = 0; keyIndex < keyNumber; keyIndex++) {
						int columnIndex = keyIndex + 1;
						insertStatement.setObject(columnIndex, cypherRecord.get(keyIndex).asObject());
					}
					insertStatement.addBatch();
				}
				insertStatement.executeBatch();
			}
		}
	}
}

File: cypher-to-sql/src/main/scala/ingraph/compiler/sql/ExportStep.scala
package ingraph.compiler.sql

import java.sql.Connection

import ingraph.compiler.sql.Util.withResources
import ingraph.compiler.sql.driver.ValueJsonConversion
import org.neo4j.driver.v1.{Session, Value}

import scala.collection.JavaConverters._

class ExportStep(val exportCypherQuery: String, val tableName: String) {

  def exportToTable(cypherSession: Session, sqlConnection: Connection): Unit = {
    val cypherResult = cypherSession.run(exportCypherQuery)

    val keysInRecord = cypherResult.keys.size
    val valueParameters = cypherResult.keys.asScala
      .map(key =>
        if (key == "value")
          "?::jsonb"
        else
          "?")
      .mkString(", ")
    val insertQueryString = s"INSERT INTO $tableName VALUES ($valueParameters)"

    withResources(sqlConnection.prepareStatement(insertQueryString))(insertStatement => {
      for (cypherRecord <- cypherResult.asScala) {
        for (keyIndex <- 0 until keysInRecord) {
          val columnIndex = keyIndex + 1
          val cypherValue = cypherRecord.get(keyIndex)
          val cypherValueObject = cypherValue.asObject

          val value =
            if (cypherRecord.keys().get(keyIndex) == "value")
              ValueJsonConversion.gson.toJson(cypherValue, classOf[Value])
            else
              cypherValueObject

          insertStatement.setObject(columnIndex, value)
        }
        insertStatement.addBatch()
      }

      insertStatement.executeBatch()
    })
  }
}
